crates/notion-core/src/catalog.rs

```diff
@@ -153,74 +213,49 @@ impl NotionFail for NoNodeVersionFoundError {
+struct NoYarnVersionFoundError {
+    matching: VersionReq,
 }
-
-/// Reads a public index from the Node cache, if it exists and hasn't expired.
-fn read_cached_opt() -> Fallible<Option<serial::index::Index>> {
-    let expiry: Option<String> = read_file_opt(&path::node_index_expiry_file()?).unknown()?;
-
-    if let Some(string) = expiry {
-        let expiry_date: HttpDate = HttpDate::from_str(&string).unknown()?;
-        let current_date: HttpDate = HttpDate::from(SystemTime::now());
-
-        if current_date < expiry_date {
-            let cached: Option<String> = read_file_opt(&path::node_index_file()?).unknown()?;
-
-            if let Some(string) = cached {
-                return Ok(serde_json::de::from_str(&string).unknown()?);
-            }
-        }
+impl NotionFail for NoYarnVersionFoundError {
+    fn is_user_friendly(&self) -> bool {
+        true
     }
-
-    Ok(None)
-}
-
-/// Get the cache max-age of an HTTP reponse.
-fn max_age(response: &reqwest::Response) -> u32 {
-    if let Some(cache_control_header) = response.headers().get::<CacheControl>() {
-        for cache_directive in cache_control_header.iter() {
-            if let CacheDirective::MaxAge(max_age) = cache_directive {
-                return *max_age;
-            }
-        }
+    fn exit_code(&self) -> i32 {
+        100
     }
-
-    // Default to four hours.
-    4 * 60 * 60
 }
 
-impl NodeCatalog {
-    /// Tests whether this Node catalog contains the specified Node version.
+impl<I:Install> Collection<I> {
+    /// Tests whether this Collection contains the specified Tool version.
     pub fn contains(&self, version: &Version) -> bool {
         self.versions.contains(version)
     }
+}
 
+pub trait Resolve<I:Install> {
     /// Resolves the specified semantic versioning requirements from a remote distributor.
-    fn resolve_remote(&self, matching: &VersionReq, config: &Config) -> Fallible<Installer> {
-        match config.node {
-            Some(NodeConfig {
-                resolve: Some(ref plugin),
-                ..
-            }) => plugin.resolve(matching),
+    fn resolve_remote(&self, matching: &VersionReq, config: Option<&ToolConfig<I>>) -> Fallible<I> {
+        match config {
+            Some(ToolConfig {
+                 resolve: Some(ref plugin),
+                 ..
+             }) => plugin.resolve(matching),
             _ => self.resolve_public(matching),
         }
     }
 
-    /// Resolves the specified semantic versioning requirements from the public distributor (`https://nodejs.org`).
-    fn resolve_public(&self, matching: &VersionReq) -> Fallible<Installer> {
+    /// Resolves the specified semantic versioning requirements from the public distributor (e.g. `https://nodejs.org`).
+    fn resolve_public(&self, matching: &VersionReq) -> Fallible<I>;
+
+    /// Resolves the specified semantic versioning requirements from the local catalog.
+    fn resolve_local(&self, req: &VersionReq) -> Option<Version>;
```

Comment:
Isn't the behavior of `resolve_local` identical for all implementations? If so, you should be able to hoist the definition here and avoid the duplication between the `YarnCollection` and `NodeCollection` implementations.

Reply:
Good point, and makes total sense, but as we spoke offline I couldn't simply move the method to the `Resole` trait since `resolve_local` has reference to `self.versions` which is a field on the `Collection` struct. I'd love to get your help and learn how can I abstract that.